draft-gpew-priv-ppm.md

```diff
@@ -969,11 +960,27 @@ prevent such replay attacks, this specification requires the aggregators to
 detect and filter out replayed reports.
 
 To detect replay attacks, each aggregator keeps track of the set of nonces
-pertaining to reports that were previously aggregated for a given task. If the
-leader receives a report from a client whose nonce is in this set, it simply
-ignores it. A helper who receives an encrypted input share whose nonce is
-in this set replies to the leader with an error as described in
-{{aggregate-request}}.
+pertaining to reports that were previously aggregated for a given task. Before
+processing a report, an aggregator first checks if the report pertains to a
+batch that has already been "collected". This is the case if the aggregator is
+the leader and has completed a collect request for a batch containing the report
+(see {{pa-collect}}) or if the aggregator is a helper and has completed an
+aggregate-share request for a batch containing the report (see
+{{aggregate-share-request}}). It also checks if the report has been
+"aggregated". This is the case if a report with the same nonce was previously
+used in an aggregate request.
+
+* If the report has not yet been aggregated but is contained by a batch that has
+  been collected, then the aggregator MUST ignore it. This prevents additional
+  reports from being aggregated after its batch has already been collected.
+
+* If the report has been aggregated but has not been collected, then the
+  aggregator MUST ignore it. This prevents a report from being used more than
+  once in a batch.
```

Comment:
I think but am not certain that this needs to reference the task's `max_batch_lifetime` rather than allowing exactly once, because successive collector queries will change the aggregation parameter (e.g., different prefixes in `poplar1`).

Reply:
I don't think want to change the batch after it has been collected with *any* aggregation parameter. Imagine, for example, that the collector requests the aggregate for candidate prefixes ["ab"]. Then add another measurement and re-run with candidate prefixes ["abc"]. Then I know immediately whether the additional measurement was prefixed by "abc".

Reply:
Then I think there's something I must not be understanding about how `max_batch_lifetime` works, or how PPM is meant to solve t-heavy-hitters, which I thought involved the collector sending a series of collect requests with longer candidate prefixes each time. But that doesn't need to be solved in this PR.

Reply:
You understand correctly how PPM is supposed to solve heavy hitters. All `max_batch_lifetime` is supposed to say is that you can't make any more collect requests than `max_batch_lifetime`. Otherwise it doesn't make any restrictions on the aggregation parameter. In particular, Poplar can be implemented as you describe.

All that this line is saying is that you can't change the set of reports in a batch after you've collected it. Otherwise, this could lead to a privacy violation, regardless of the VDAF in use.